File: sample-conference-videochat/sample-conference-videochat/Components/Users/Classes/MainDataSource/MainDataSource.m
//
//  MainDataSource.m
//  sample-multiconference-videochat
//
//  Copyright (c) 2017 QuickBlox. All rights reserved.
//

#import "MainDataSource.h"

@interface MainDataSource () {
    
    SEL _sortSelector;
}

@property (nonatomic, readwrite) NSArray *selectedObjects;

@end

@implementation MainDataSource

// MARK: Construction

- (instancetype)initWithSortSelector:(SEL)sortSelector {
    
    self = [super init];
    if (self != nil) {
        
        _sortSelector = sortSelector;
        _objects = [[NSArray alloc] init];
        _selectedObjects = [[NSArray alloc] init];
    }
    
    return self;
}

// MARK: Public

- (void)setObjects:(NSArray *)objects {
    
    if (![_objects isEqualToArray:objects]) {
        
        _objects = [self sortObjects:objects];
        
        NSMutableArray *mutableSelectedObjects = [_selectedObjects mutableCopy];
        for (id obj in self.selectedObjects) {
            
            if (![_objects containsObject:obj]) {
                [mutableSelectedObjects removeObject:obj];
            }
        }
        _selectedObjects = [mutableSelectedObjects copy];
    }
}

- (NSArray *)selectedObjects {
    return [_selectedObjects copy];
}

- (void)selectObjectAtIndexPath:(NSIndexPath *)indexPath {
    
    id obj = _objects[indexPath.row];
    
    NSMutableArray *mutableSelectedObjects = [_selectedObjects mutableCopy];
    if ([_selectedObjects containsObject:obj]) {
        [mutableSelectedObjects removeObject:obj];
    }
    else {
        [mutableSelectedObjects addObject:obj];
    }
    _selectedObjects = [mutableSelectedObjects copy];
}

- (void)deselectAllObjects {
    _selectedObjects = [[NSArray alloc] init];
}

// MARK: Private

- (NSArray *)sortObjects:(NSArray *)objects {
    
    // Create sort Descriptor
    NSSortDescriptor *objectsSortDescriptor =
    [[NSSortDescriptor alloc] initWithKey:NSStringFromSelector(_sortSelector)
                                ascending:NO];
    
    NSArray *sortedObjects = [objects sortedArrayUsingDescriptors:@[objectsSortDescriptor]];
    
    return sortedObjects;
}

// MARK: UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return _objects.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSAssert(nil, @"Required to be implemented by subclass.");
    return nil;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Users/Classes/MainDataSource/MainDataSource.swift
//
//  MainDataSource.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 11.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

class MainDataSource<T: Any> : NSObject, UITableViewDataSource where T: Equatable {
    
    var objects = [T]()
    var selectedObjects = [T]()
    
    private var sortSelector: Selector? {
        didSet {
            debugPrint("sortSelector did set \(String(describing: sortSelector))")
        }
    }
    
    // MARK: Construction
    init(sortSelector: Selector) {
        self.sortSelector = sortSelector
    }
    
    // MARK: Public
    func updateObjects(_ objects: [T]) {
        self.objects = sortObjects(objects)
        selectedObjects = selectedObjects.filter({ objects.contains($0) })
    }
    
    func addObjects(_ objects: [T]) {
        updateObjects(self.objects + objects)
    }
    
    func selectObject(at indexPath: IndexPath?) {
        guard let indexPath = indexPath else { return }
        let object = objects[indexPath.row]
            if selectedObjects.contains(object) {
                selectedObjects.removeAll(where: { element in element == object })
            } else {
                selectedObjects.append(object)
            }
    }
    
    func deselectAllObjects() {
        selectedObjects = [T]()
    }
    
    // MARK: Private
    func sortObjects(_ objects: [T]) -> [T] {
        
        // Create sort Descriptor
        guard let sortSelector = self.sortSelector else {
            return objects
        }
        let key = NSStringFromSelector(sortSelector)
        let objectsSortDescriptor = NSSortDescriptor(key: key, ascending: false)
        guard let sortedObjects = (objects as NSArray).sortedArray(using: [objectsSortDescriptor])
            as? [T] else {return objects}
        return sortedObjects
    }
    
    // MARK: UITableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return objects.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        assert(false, "Required to be implemented by subclass.")
        let cell = UITableViewCell()
        return cell
    }
}
